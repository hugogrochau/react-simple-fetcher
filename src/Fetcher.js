import React, { PureComponent } from 'react';

export default class Fetcher extends PureComponent {
  constructor (props) {
    super(props);

    this.state = {
      fetching: true,
      data: null
    }
  }

  async componentDidMount () {
    const { handler } = this.props;

    if (typeof handler === 'function') {
      const data = await handler();

      this.setState({
        fetching: false,
        data
      });
    }
  }

  render () {
    const { component, render, children } = this.props;
    const props = this.state;

    if (component) {
      return React.createElement(component, props);
    }

    if (render) {
      return render(props);
    }

    if (typeof children === 'function') {
      return children(props);
    }

    if (children && React.Children.count(children) > 0) {
      return React.Children.only(children);
    }

    return null;
  }
}
