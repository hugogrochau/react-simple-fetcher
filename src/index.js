export { default } from './Fetcher';
export { default as connectFetcher } from './connectFetcher';
