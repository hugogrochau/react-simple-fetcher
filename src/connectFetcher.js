import React from 'react';
import Fetcher from './Fetcher';

const connectFetcher = fetch => Component => {
  const C = props => {
    return (
      <Fetcher fetch={fetch} render={fetcherProps => (
        <Component {...props} {...fetcherProps} />
      )} />
    );
  }

  C.displayName = `connectFetcher(${Component.displayName || Component.name})`;

  return C;
}

export default connectFetcher;
