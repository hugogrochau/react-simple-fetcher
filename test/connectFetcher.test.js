import React from 'react';
import renderer from 'react-test-renderer';
import { connectFetcher } from '../src';


describe('connectFetcher()(Component)', () => {
  it('should render', () => {
    const Component = props => <div>Component</div>;
    const ConnectFetcher = connectFetcher(null)(Component);

    const tree = renderer.create(
      <ConnectFetcher />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should receive { fetching, data, fake } props', () => {
    let actual = null;
    const Component = props => (actual = props) && null;
    const ConnectFetcher = connectFetcher(null)(Component);

    renderer.create(
      <ConnectFetcher fake="prop" />
    );

    expect(actual.fetching).toBe(true);
    expect(actual.data).toBe(null);
    expect(actual.fake).toBe('prop');
  });
});

describe('connectFetcher(fetch)(MockedComponent)', () => {
  const mockFetch = () => new Promise((resolve) => {
    setTimeout(() => {
      const data = {
        fake: 'data'
      };

      resolve(data);
    }, 300);
  });

  it('should receive { fetching, data } after fetch', async () => {
    let actual = null;
    const Component = props => (actual = props) && null;
    const ConnectFetcher = connectFetcher(mockFetch)(Component);

    renderer.create(
      <ConnectFetcher />
    );

    await mockFetch();

    expect(actual.fetching).toBe(false);
    expect(actual.data).toEqual({
      fake: 'data'
    });
  });
});
