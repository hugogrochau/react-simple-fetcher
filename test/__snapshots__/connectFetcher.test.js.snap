// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectFetcher()(Component) should render 1`] = `
<div>
  Component
</div>
`;
